File: python/mirage/commands/builtin/gdocs/_provision.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.gdocs import GDocsAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.gdocs.stat import stat as gdocs_stat
from mirage.provision.types import Precision, ProvisionResult
from mirage.types import PathSpec


async def _resolve_sizes(
    accessor: GDocsAccessor,
    paths: list[PathSpec],
    index: IndexCacheStore | None,
) -> tuple[list[tuple[str, int]], int]:
    """Walk paths, return (path, size) pairs. Self-heals via stat fallback."""
    resolved: list[tuple[str, int]] = []
    missing = 0
    for p in paths:
        path_str = p.original if isinstance(p, PathSpec) else p
        size = None
        if index is not None:
            lookup = await index.get(path_str)
            if lookup.entry is not None:
                size = lookup.entry.size
        if size is None:
            try:
                file_stat = await gdocs_stat(accessor, p, index)
                size = file_stat.size
            except (FileNotFoundError, ValueError):
                pass
        if size is not None:
            resolved.append((path_str, size))
        else:
            missing += 1
    return resolved, missing


async def file_read_provision(
    accessor: GDocsAccessor,
    paths: list[PathSpec],
    command: str = "",
    index: IndexCacheStore | None = None,
    **_extra: object,
) -> ProvisionResult:
    """Cost estimate for full file reads (cat, wc) over Google Docs."""
    if not paths:
        return ProvisionResult(command=command, precision=Precision.UNKNOWN)
    if index is None:
        index = index
    resolved, missing = await _resolve_sizes(accessor, paths, index)
    if missing > 0 or not resolved:
        return ProvisionResult(command=command, precision=Precision.UNKNOWN)
    total = sum(size for _, size in resolved)
    return ProvisionResult(
        command=command,
        network_read_low=total,
        network_read_high=total,
        read_ops=len(resolved),
        precision=Precision.EXACT,
    )


async def metadata_provision(command: str = "",
                             **_extra: object) -> ProvisionResult:
    """Cost estimate for metadata-only ops over Google Docs."""
    return ProvisionResult(
        command=command,
        network_read_low=0,
        network_read_high=0,
        read_ops=0,
        precision=Precision.EXACT,
    )

File: typescript/packages/core/src/commands/builtin/databricks_volume/provision.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { Accessor } from '../../../accessor/base.ts'
import type { IndexCacheStore } from '../../../cache/index/store.ts'
import type { PathSpec } from '../../../types.ts'
import { Precision, ProvisionResult } from '../../../provision/types.ts'
import type { DatabricksVolumeAccessor } from '../../../accessor/databricks_volume.ts'
import type { CommandOpts } from '../../config.ts'
import { stat as dbxStat } from '../../../core/databricks_volume/stat.ts'

interface ResolvedSizes {
  resolved: [string, number][]
  missing: number
}

async function resolveSizes(
  accessor: DatabricksVolumeAccessor,
  paths: readonly PathSpec[],
  index: IndexCacheStore | undefined,
): Promise<ResolvedSizes> {
  const resolved: [string, number][] = []
  let missing = 0
  for (const p of paths) {
    const pathStr = p.original
    let size: number | null = null
    if (index !== undefined) {
      const lookup = await index.get(pathStr)
      if (lookup.entry !== undefined && lookup.entry !== null) {
        size = lookup.entry.size
      }
    }
    if (size === null) {
      try {
        const fileStat = await dbxStat(accessor, p)
        size = fileStat.size
      } catch {
        // fall through — size stays null
      }
    }
    if (size !== null) {
      resolved.push([pathStr, size])
    } else {
      missing += 1
    }
  }
  return { resolved, missing }
}

function parseNumFlag(value: string | boolean | undefined): number | null {
  if (typeof value !== 'string') return null
  const n = Number.parseInt(value, 10)
  return Number.isFinite(n) ? n : null
}

export async function fileReadProvision(
  accessor: Accessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<ProvisionResult> {
  if (paths.length === 0) {
    return new ProvisionResult({ precision: Precision.UNKNOWN })
  }
  const index = opts.index ?? undefined
  const { resolved, missing } = await resolveSizes(
    accessor as DatabricksVolumeAccessor,
    paths,
    index,
  )
  if (missing > 0 || resolved.length === 0) {
    return new ProvisionResult({ precision: Precision.UNKNOWN })
  }
  const total = resolved.reduce((sum, [, size]) => sum + size, 0)
  return new ProvisionResult({
    networkReadLow: total,
    networkReadHigh: total,
    readOps: resolved.length,
    precision: Precision.EXACT,
  })
}

export async function headTailProvision(
  accessor: Accessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<ProvisionResult> {
  if (paths.length === 0) {
    return new ProvisionResult({ precision: Precision.UNKNOWN })
  }
  const index = opts.index ?? undefined
  const { resolved, missing } = await resolveSizes(
    accessor as DatabricksVolumeAccessor,
    paths,
    index,
  )
  if (missing > 0 || resolved.length === 0) {
    return new ProvisionResult({ precision: Precision.UNKNOWN })
  }
  const c = parseNumFlag(opts.flags.c)
  if (c !== null) {
    const total = resolved.reduce((sum, [, size]) => sum + Math.min(c, size), 0)
    return new ProvisionResult({
      networkReadLow: total,
      networkReadHigh: total,
      readOps: resolved.length,
      precision: Precision.EXACT,
    })
  }
  const full = resolved.reduce((sum, [, size]) => sum + size, 0)
  return new ProvisionResult({
    networkReadLow: 0,
    networkReadHigh: full,
    readOps: resolved.length,
    precision: Precision.RANGE,
  })
}

export function metadataProvision(
  _accessor: Accessor,
  paths: PathSpec[],
  _texts: string[],
  _opts: CommandOpts,
): ProvisionResult {
  const n = Math.max(1, paths.length > 0 ? paths.length : 1)
  return new ProvisionResult({
    networkReadLow: 0,
    networkReadHigh: 0,
    readOps: n,
    precision: Precision.EXACT,
  })
}
